### ISSUE TITLE
HorizontalPodAutoscaler v2beta2 bug

### ISSUE THREAD

**Issue by L-maple** [kubernetes/kubernetes#99349]
#### What happened:
When I use HPA v2beta2 with custom metric, I found the hpa automatically scaleup the pods out of control and the described reason is "All metrics below targets". The HPA description is as follows:
```
(base) [root@master pv_collector]# k describe hpa -n monitoring  hdfs-datanode-custom 
Name:                                                                     hdfs-datanode-custom
Namespace:                                                                monitoring
Labels:                                                                   <none>
Annotations:                                                              <none>
CreationTimestamp:                                                        Tue, 23 Feb 2021 02:54:59 -0500
Reference:                                                                StatefulSet/hdfs-datanode
Metrics:                                                                  ( current / target )
  "whether_add_pod" on Service/hdfs-hpa-exporter-service (target value):  0 / 100
Min replicas:                                                             3
Max replicas:                                                             8
Behavior:
  Scale Up:
    Stabilization Window: 30 seconds
    Select Policy: Min
    Policies:
      - Type: Pods  Value: 1  Period: 120 seconds
  Scale Down:
    Select Policy: Disabled
    Policies:
      - Type: Percent  Value: 100  Period: 15 seconds
StatefulSet pods:      5 current / 6 desired
Conditions:
  Type            Status  Reason            Message
  ----            ------  ------            -------
  AbleToScale     True    SucceededRescale  the HPA controller was able to update the target scale to 6
  ScalingActive   True    ValidMetricFound  the HPA was able to successfully calculate a replica count from Service metric whether_add_pod
  ScalingLimited  True    ScaleUpLimit      the desired replica count is increasing faster than the maximum scale rate
Events:
  Type    Reason             Age   From                       Message
  ----    ------             ----  ----                       -------
  Normal  SuccessfulRescale  4m6s  horizontal-pod-autoscaler  New size: 4; reason: Service metric whether_add_pod above target
  Normal  SuccessfulRescale  2m3s  horizontal-pod-autoscaler  New size: 5; reason: All metrics below target
  Normal  SuccessfulRescale  0s    horizontal-pod-autoscaler  New size: 6; reason: All metrics below target
```

When I check the custom metric,  I find actually the metric, which is 0, is below target. 

So what's the real reason HPA scaleup??

#### What you expected to happen:
I expect when the custom metric is below target, hpa will not scaleup the pods automatically, which is really strange...

Background:
I use HPA on statefulSet, so I disable the scaledown in the spec.behavior.  When the whether_add_pod(custom_metrics) is equal to or above 100(the configured target value), the HPA would scaleup the pods. In other words, if whether_add_pod is below 100, it would not scaleup the pods.
And I hope I could control the scaleup process by whether_add_pod(custom metric).

#### How to reproduce it (as minimally and precisely as possible):
I use HPA to control the statefulset, and yy hpa configuration is as below.
When the HPA scaleup because of the custom metrics, it will scaleup automatically out of control. As the above status：
```
Events:
  Type    Reason             Age   From                       Message
  ----    ------             ----  ----                       -------
  Normal  SuccessfulRescale  4m6s  horizontal-pod-autoscaler  New size: 4; reason: Service metric whether_add_pod above target
  Normal  SuccessfulRescale  2m3s  horizontal-pod-autoscaler  New size: 5; reason: All metrics below target
  Normal  SuccessfulRescale  0s    horizontal-pod-autoscaler  New size: 6; reason: All metrics below target
```

HPA configuration：
```
apiVersion: autoscaling/v2beta2
kind: HorizontalPodAutoscaler
metadata:
  name: hdfs-datanode-custom
  namespace: monitoring
spec:
  scaleTargetRef:
    apiVersion: apps/v1
    kind: StatefulSet
    name: hdfs-datanode
  minReplicas: 3
  maxReplicas: 8
  metrics:
    -type: Object
    object:
      metric:
        name: whether_add_pod
      describedObject:
        apiVersion: v1
        kind: Service
        name: hdfs-hpa-exporter-service
      target:
        -type: Value
        value: 100
  behavior:
    scaleDown:
      selectPolicy: Disabled  # 禁止缩容
    scaleUp:
      stabilizationWindowSeconds: 30
      policies:
          -type: Pods
          value: 1
          periodSeconds: 120    # 表示策略的时间长度必须保证有效，该策略中每120s最多增加1个Pod
      selectPolicy: Min        # 通过设置Min值，HPA将选择副本数量变化最小的策略；
```

#### Anything else we need to know?:
When I check the events, I realize the HPA scale every 2 minute automatically, and the 2 minutes is what I set in the configuration. 

#### Environment:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:09:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```
- Cloud provider or hardware configuration: private cluster
- OS (e.g: `cat /etc/os-release`): 
```
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Kernel (e.g. `uname -a`):
```
Linux master 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
```
- Install tools:
kubeadm
- Network plugin and version (if this is a network-related bug):
flannel
- Others:

**Comment by JoshuaAndrew** [comment]
@L-maple  metric whether_add_pod above target, so scaleup behavior works as expected,

**Comment by L-maple** [comment]
> @L-maple metric whether_add_pod above target, so scaleup behavior works as expected,

Really thanks fot your quick tips, but I mean the following events are strange. The HPA adds pod automatically...
```
  Normal  SuccessfulRescale  2m3s  horizontal-pod-autoscaler  New size: 5; reason: All metrics below target
  Normal  SuccessfulRescale  0s    horizontal-pod-autoscaler  New size: 6; reason: All metrics below target
```

**Comment by L-maple** [comment]
@JoshuaAndrew Hi~ Do you have any other suggestions? Disabling scaledown would cause this problem?(My guess...)

**Comment by igtsekov** [comment]
I have the exact same issue. @L-maple were you able to find a workaround?

**Comment by igtsekov** [comment]
I might have found a workaround. 
In my setup, I have a global HPA scale down stabilization period of 5 minutes (300s), that we set with the `--horizontal-pod-autoscaler-downscale-stabilization` option of the [kube-control-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/). If I set [this](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#stabilization-window) 
```
behaviour:
  scaleDown:
    stabilizationWindowSeconds: <number>
```
to the HPA where `<number>` is a smaller value (I tested it with 60, 150 and 250) the HPA behaviour is as expected and the bug does NOT occur.

**Comment by druman55** [comment]
I am having the same problem.

FYI - setting `stabilizationWindowSeconds` to 299 did not resolve the issue for me.

Note that the current HPA value is 20, and target is 100.  Seemingly whenever my metrics rise above 0 the deployment scales up aggressively - sometime straight to max and sometimes it takes a few 15 second windows to reach max (controlled by the scale up policy it seems).

**kubectl version:**
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.1", GitCommit:"5e58841cce77d4bc13713ad2b91fa0d961e69192", GitTreeState:"clean", BuildDate:"2021-05-12T14:18:45Z", GoVersion:"go1.16.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}

HPA:

```Name:                                                                               action-processor-v2-highmem
Namespace:                                                                          default
Labels:                                                                             <none>
Annotations:                                                                        <none>
CreationTimestamp:                                                                  Thu, 10 Jun 2021 09:15:30 -0400
Reference:                                                                          Deployment/action-processor-v2-highmem
Metrics:                                                                            ( current / target )
  "rabbitmq_queue_messages" on Service/action-processor-v2-highmem (target value):  20 / 100
Min replicas:                                                                       2
Max replicas:                                                                       5
Behavior:
  Scale Up:
    Stabilization Window: 0 seconds
    Select Policy: Max
    Policies:
      - Type: Percent  Value: 100  Period: 15 seconds
      - Type: Pods     Value: 4    Period: 15 seconds
  Scale Down:
    Stabilization Window: 299 seconds
    Select Policy: Max
    Policies:
      - Type: Percent  Value: 100  Period: 15 seconds
Deployment pods:       2 current / 5 desired
Conditions:
  Type            Status  Reason            Message
  ----            ------  ------            -------
  AbleToScale     True    SucceededRescale  the HPA controller was able to update the target scale to 5
  ScalingActive   True    ValidMetricFound  the HPA was able to successfully calculate a replica count from Service metric rabbitmq_queue_messages
  ScalingLimited  True    TooManyReplicas   the desired replica count is more than the maximum replica count
Events:
  Type    Reason             Age   From                       Message
  ----    ------             ----  ----                       -------
  Normal  SuccessfulRescale  16s   horizontal-pod-autoscaler  New size: 2; reason: Current number of replicas below Spec.MinReplicas
  Normal  SuccessfulRescale  1s    horizontal-pod-autoscaler  New size: 5; reason: Service metric rabbitmq_queue_messages above target`

**Comment by druman55** [comment]
As additional follow up, everything about the HPA seems to be working properly as far as the behavior settings.  It's just looking at 20/100 and thinking "Yep, we should scale up, ALL the way!"

Is there a way to get more verbosity from `horizontal-pod-autoscaler` ?

**Comment by druman55** [comment]
At least on our end, the mystery has been solved!!!

When we create Deployments we do it by replacing several variables in a Deployment Template.  Most of these replacements are for the serviceName, but they also include memory, CPU reservations, etc which we manage elsewhere.

One of the replacements was missed, and all of the deployments on our cluster were being spun up looking like the same "Run or "App".

In this Deployment YAML: 
`apiVersion: apps/v1
kind: Deployment
metadata:
  name: serviceName
  namespace: default
spec:
  replicas: 1
  selector:
    matchLabels:
      app: serviceName
  strategy:
    type: RollingUpdate
  template:
    metadata:
      labels:
        run: serviceName
        app: serviceName
`
One (or possibly both) of template.metadata.labels.run and template.metadata.labels.app were remaining as serviceName - and even tho the HPA in no way references the run or app attributes, the HPA controller must use it to aggregate values together to determine something about the underlying math - and it was causing the rampant overscaling.

I adjusted our runService script to do the appropriate replacements, re-built all of the services and now the scaling is working exactly as expected.

So, check your values for these attributes and see if they are attributes you never changed from a default perhaps.
